lib/bigdecimal.rb

```diff
@@ -3,3 +3,27 @@
 else
   require 'bigdecimal.so'
 end
+
+class BigDecimal
+
+  # Returns the square root of the value.
+  #
+  # Result has at least prec significant digits.
+  #
+  def sqrt(prec)
+    if infinite? == 1
+      exception_mode = BigDecimal.mode(BigDecimal::EXCEPTION_ALL)
+      raise FloatDomainError, "Computation results in 'Infinity'" if exception_mode.anybits?(BigDecimal::EXCEPTION_INFINITY)
+      return INFINITY
+    end
+    raise ArgumentError, 'negative precision' if prec < 0
+    raise FloatDomainError, 'sqrt of negative value' if self < 0
+    raise FloatDomainError, "sqrt of 'NaN'(Not a Number)" if nan?
+
+    prec = [prec, n_significant_digits].max
+    ex = prec + BigDecimal.double_fig - exponent / 2
+    base = BigDecimal(10) ** ex
+    sqrt = Integer.sqrt(self * base * base)
```

Comment:
How about adding the new methods for decimal shift operation to calculate the multiplication by $10^n$ faster?

Reply:
Cost of multiplication/division with 10**n is not dominant in normal path, but in the fast path such as  `BigDecimal('16e1000000').sqrt(1000000)`, the improvement of using decimal shift operation is significant.

Reply:
How much does the calculation of `10**n` cost?  I think we can further reduce the total cost by following two ways:

1. Introduce the decimal shift operation I mentioned previously.
2. Use `BigDecimal("10e#{ex}")` instead of using `BigDecimal#**` for calculating `10**n`.

We reduce the cost of `self * base * base` by the former, and the cost of calculating `base` by the latter.

Instead of the latter way, implementing the fast pass for `10**n` in `VpPowerByInt` is ok to me.

Reply:
| scenario | decimal_shift | `"1e{prec}"` | `ten**prec` |
| --- | ------------ | ------------- | ----------- |
| prec=10, 100000.times | 0.149624s | 0.244520s | 0.245961s |
| prec=100, 100000.times | 0.442031s | 0.481656s | 0.605150s |
| prec=1000, 100000.times | 2.846068s | 2.963888s | 3.544875s |
| prec=10000, 10000.times | 3.441482s | 3.443580s | 3.844375s |

Comparision of:
```ruby
BigDecimal(Integer.sqrt(BigDecimal(2)._decimal_shift(2*prec)))._decimal_shift(-prec)

Integer.sqrt(BigDecimal(2)*BigDecimal("1e#{2*prec}"))*BigDecimal("1e-#{prec}")

base = BigDecimal(10)**prec; Integer.sqrt(BigDecimal(2)*base*base)/base
```

Reply:
Updated the implementation c804b09e27e88bc99fb108376a6f2845d4f46539 with `* BigDecimal("1e#{2 * ex}")` and `* BigDecimal("1e#{-ex}")` (this is 4 times faster than division)